RFC-0008-plan-checker.md

```diff
@@ -0,0 +1,137 @@
+# **RFC0 for Presto**
+
+See [CONTRIBUTING.md](CONTRIBUTING.md) for instructions on creating your RFC and the process surrounding it.
+
+## [Title]
+
+Proposers
+
+* Tim Meehan
+* Bryan Cutler
+* Rebecca Schlussel
+
+## [Related Issues]
+
+* RFC-0003
+
+## Summary
+
+Add a new SPI to integrate a custom plan checker, and add a plugin to use the Presto sidecar to check if a Presto plan can be
+successfully translated into a Velox plan.
+
+## Background
+
+The optimizer makes decisions in part based on the capabilities of the underlying evaluation engine.  With the Presto evaluation
+engine being migrated to Velox while concurrently supporting the Presto Java evaluation engine, there's now differences between
+what is supported between both evaluation engines.  Because of these underlying differences, the optimizer may generate plans that
+can't be executed in C++ clusters, or likewise, a Presto Java cluster could be misconfigured to generate plans that only work with
+C++ clusters.
+
+If a plan is generated that can't be executed in the cluster, the query will fail with an error message during the query execution
+phase.  This is a poor user experience, as the user has to wait for the query to fail before they can take corrective action, and
+this might have occurred after lengthy queueing.  Additionally, this failure would occur after worker resources have already been
+allocated to the query, which is wasteful.  This RFC proposes a plan checker that can be run before the query is executed to ensure
+a quick validation of the plan.
+
+### [Optional] Goals
+
+* Provide a mechanism to validate Presto to Velox plan conversion during planning phase
+* Add an SPI to allow custom validators to be added to suit individual business needs
+
+### [Optional] Non-goals
+
+* Ensure all Velox plans are executable in a Velox cluster--many checks are done at runtime and may not be caught by the plan checker
+
+## Proposed Implementation
+
+The Presto runtime centralizes plan validation logic into the `PlanChecker` class.  There exist three phases to this class: 
+
+* `validateIntermediatePlan`
+* `validateFinalPlan`
+* `validateFragmentedPlan`
+
+It is generally useful to allow this class to be configured with an SPI that allows for custom plan validation logic to be added.
+For example, a business may decide that a certain type is not allowed, or add a check to ensure that plans that are overly
+complicated are killed.
+
+An SPI will be added that will add more checks to the `PlanChecker` class.  The SPI will contain a field indicating which phase
```

Comment:
Just to clarify the SPI will extend the built-in `PlanChecker` by allowing additional `Checker`s for each of the plan phases.